File: utm-seclib-node/src/ussp-pki/pki-certs.ts
import { CreateLogger } from '../utils';
import * as crypto from 'crypto';

const logs = CreateLogger();

export interface CertificateAttributes {
    certificateSigningRequest: crypto.X509Certificate;
    signerPrivateKey: crypto.KeyObject;
    parentCert?: crypto.X509Certificate;
    isCA: boolean;
    validForDays?: number;
}

export class DistinguishedName {
    constructor(
        public readonly commonName: string,
        public readonly organization: string,
        public readonly organizationUnit: string,
        public readonly country: string,
        public readonly locality: string
    ) {
        this.validateFields();
    }

    private validateFields(): void {
        const requiredFields = [
            { value: this.commonName, name: 'Common name' },
            { value: this.organization, name: 'Organization' },
            { value: this.organizationUnit, name: 'Organization unit' },
            { value: this.country, name: 'Country' },
            { value: this.locality, name: 'Locality' }
        ];

        for (const { value, name } of requiredFields) {
            if (!value?.trim()) {
                const message = `${name} cannot be empty`;
                logs.error(message);
                throw new Error(message);
            }
        }
    }
}

interface X509Subject {
    readonly CN: string;
    readonly O: string;
    readonly OU: string;
    readonly C: string;
    readonly L: string;
}

export function CreateX509Subject(distinguishedName: DistinguishedName): string {
    try {
        const subject = mapToX509Subject(distinguishedName);
        checkX509SubjectRequiredFields(subject);
        return formatX509Subject(subject);
    } catch (error) {
        const message = error instanceof Error ? error.message : 'An unknown error occurred';
        logs.error(`Failed to create X.509 subject: ${message}`);
        throw new Error(message);
    }
}

function mapToX509Subject(dn: DistinguishedName): X509Subject {
    return {
        CN: dn.commonName,
        O: dn.organization,
        OU: dn.organizationUnit,
        C: dn.country,
        L: dn.locality,
    };
}

function checkX509SubjectRequiredFields(subject: X509Subject): void {
    const requiredFields: Array<{ field: keyof X509Subject; name: string }> = [
        { field: 'CN', name: 'Common name' },
        { field: 'O', name: 'Organization' },
        { field: 'OU', name: 'Organization unit' },
        { field: 'C', name: 'Country' },
        { field: 'L', name: 'Locality' }
    ];

    for (const { field, name } of requiredFields) {
        if (!subject[field]?.trim()) {
            const message = `${name} cannot be empty`;
            logs.error(message);
            throw new Error(message);
        }
    }
}

function formatX509Subject(subject: X509Subject): string {
    const fields: Array<keyof X509Subject> = ['CN', 'OU', 'O', 'L', 'C'];
    return fields.map(field => `${field}=${subject[field]}`).join(',');
}

File: utm-seclib-go/pkg/ussp-pki/pki-certs.go
package ussp_pki

import (
	"crypto"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"crypto/x509/pkix"
	"fmt"
	"math/big"
	"time"

	"github.com/royzah/utm-sec-lib/utm-seclib-go/pkg/utils"
)

type CertificateAttributes struct {
	CertificateSigningRequest *x509.CertificateRequest
	SignerPrivateKey          crypto.Signer
	ParentCert                *x509.Certificate
	IsCA                      bool
	ValidForDays              int
}

type DistinguishedName struct {
	CommonName       string
	Organization     string
	OrganizationUnit string
	Country          string
	Locality         string
}

func CreateX509Subject(distinguishedName DistinguishedName) (subject *pkix.Name, err error) {
	subject = &pkix.Name{
		CommonName:         distinguishedName.CommonName,
		Organization:       []string{distinguishedName.Organization},
		OrganizationalUnit: []string{distinguishedName.OrganizationUnit},
		Country:            []string{distinguishedName.Country},
		Locality:           []string{distinguishedName.Locality},
	}

	err = checkX509SubjectRequiredFields(subject)
	if err != nil {
		subject = nil
	}

	return subject, err
}

func checkX509SubjectRequiredFields(subject *pkix.Name) error {
	if subject.CommonName == "" {
		return fmt.Errorf("common name is empty")
	} else if subject.Organization[0] == "" {
		return fmt.Errorf("organization is empty")
	} else if subject.OrganizationalUnit[0] == "" {
		return fmt.Errorf("organization unit is empty")
	} else if subject.Country[0] == "" {
		return fmt.Errorf("country is empty")
	}

	return nil
}

func CreateCSR(certPrivateKey crypto.Signer, caSubject pkix.Name) (*x509.CertificateRequest, error) {

	certificateTemplate := &x509.CertificateRequest{
		Subject:            caSubject,
		PublicKey:          certPrivateKey.Public(),
		SignatureAlgorithm: getSignatureAlgorithmForPrivateKey(certPrivateKey),
	}

	certBytes, err := x509.CreateCertificateRequest(rand.Reader, certificateTemplate, certPrivateKey)
	if err != nil {
		return nil, err
	}

	return x509.ParseCertificateRequest(certBytes)
}

func getSignatureAlgorithmForPrivateKey(privateKey crypto.Signer) x509.SignatureAlgorithm {
	switch privateKey.(type) {
	case *rsa.PrivateKey:
		return x509.SHA512WithRSA
	case *ecdsa.PrivateKey:
		return x509.ECDSAWithSHA512
	case *ed25519.PrivateKey:
		return x509.PureEd25519
	default:
		return x509.UnknownSignatureAlgorithm
	}

}

func CreateX509Cert(certAttrs CertificateAttributes) (*x509.Certificate, error) {
	var certTemplate x509.Certificate

	csr := certAttrs.CertificateSigningRequest
	parentCert := certAttrs.ParentCert
	signerPrivateKey := certAttrs.SignerPrivateKey
	validForDays := certAttrs.ValidForDays

	if signerPrivateKey == nil || csr == nil {
		return nil, fmt.Errorf("missing required parameters")
	}

	if certAttrs.IsCA {
		certTemplate = getCACertificateTemplate(*big.NewInt(1), csr.Subject)
	} else {
		certTemplate = getGenericCertificateTemplate(*big.NewInt(1), csr.Subject)
	}

	if validForDays > 0 {
		certTemplate.NotAfter = time.Now().AddDate(0, 0, validForDays)
	}

	if isRootCertificate(parentCert) {
		parentCert = &certTemplate
	}

	certBytes, err := x509.CreateCertificate(rand.Reader, &certTemplate, parentCert, csr.PublicKey, signerPrivateKey)
	if err != nil {
		return nil, err
	}

	return x509.ParseCertificate(certBytes)
}

func getGenericCertificateTemplate(serialNumber big.Int, subjectName pkix.Name) x509.Certificate {
	return x509.Certificate{
		SerialNumber: &serialNumber,
		Subject:      subjectName,
		NotBefore:    time.Now(),
	}
}

func getCACertificateTemplate(serialNumber big.Int, subjectName pkix.Name) x509.Certificate {

	certTemplate := getGenericCertificateTemplate(serialNumber, subjectName)

	certTemplate.KeyUsage = x509.KeyUsageCertSign | x509.KeyUsageCRLSign
	certTemplate.IsCA = true
	certTemplate.BasicConstraintsValid = true
	certTemplate.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth}
	return certTemplate
}

func isRootCertificate(parentCert *x509.Certificate) bool {
	return parentCert == nil
}

func LoadCertFromFile(certFilePath string) (*x509.Certificate, error) {

	if certFilePath == "" {
		return nil, fmt.Errorf("empty file path given")
	}

	fileBytes, err := utils.LoadBytesFromFile(certFilePath) // #nosec G304
	if err != nil {
		return nil, err
	}

	return x509.ParseCertificate(fileBytes)
}

func SaveCertToFile(cert *x509.Certificate, certFilePath string) error {

	if cert == nil {
		return fmt.Errorf("cert must be valid x509 Certificate")
	}

	existingCert, _ := LoadCertFromFile(certFilePath)

	if existingCert != nil {
		return fmt.Errorf("certificate already exists")
	}

	certBytes := cert.Raw

	byteCount, saveBytesErr := utils.SaveBytesToFile(certBytes, certFilePath)
	if saveBytesErr != nil || byteCount == 0 {
		return fmt.Errorf("error saving certificate to file: %v", saveBytesErr)
	}

	return nil
}
